src/main/java/io/vertx/core/streams/impl/OutboundWriteQueue.java

```diff
@@ -0,0 +1,361 @@
+
+  // Immutable
+
+  private final Predicate<E> consumer;
+  private final int highWaterMark;
+  private final int lowWaterMark;
+
+  // Concurrent part accessed by any producer thread
+
+  private final Queue<E> queue = PlatformDependent.newMpscQueue();
+  private final AtomicInteger wip = new AtomicInteger();
+
+  // Consumer thread only
+
+  // The element refused by the consumer (null <=> overflow)
+  private E overflow;
+  // The number of times the queue was observed to be unwritable
+  private int writeQueueFull;
+
+  /**
+   * Create a new instance.
+   *
+   * @param consumer the predicate accepting the elements
+   * @throws NullPointerException if consumer is null
+   */
+  public OutboundWriteQueue(Predicate<E> consumer) {
+    this(consumer, DEFAULT_LOW_WATER_MARK, DEFAULT_HIGH_WATER_MARK);
+  }
+
+  /**
+   * Create a new instance.
+   *
+   * @param consumer the predicate accepting the elements
+   * @param lowWaterMark the low-water mark, must be zero or positive
+   * @param highWaterMark the high-water mark, must be greater than the low-water mark
+   *
+   * @throws NullPointerException if consumer is null
+   * @throws IllegalArgumentException if any mark violates the condition
+   */
+  public OutboundWriteQueue(Predicate<E> consumer, int lowWaterMark, int highWaterMark) {
+    Arguments.require(lowWaterMark >= 0, "The low-water mark must be > 0");
+    Arguments.require(lowWaterMark <= highWaterMark, "The high-water mark must greater or equals to the low-water mark");
+    this.consumer = Objects.requireNonNull(consumer, "Consumer must be not null");
+    this.lowWaterMark = lowWaterMark;
+    this.highWaterMark = highWaterMark;
+  }
+
+  /**
+   * @return the queue high-water mark
+   */
+  public int highWaterMark() {
+    return highWaterMark;
+  }
+
+  /**
+   * @return the queue low-water mark
+   */
+  public int lowWaterMark() {
+    return lowWaterMark;
+  }
+
+  /**
+   * Let the consumer thread add the {@code element} to the queue.
+   *
+   * A set of flags is returned
+   * <ul>
+   *   <li>When {@link #QUEUE_UNWRITABLE_MASK} is set, the queue is writable and new elements can be added to the queue,
+   *   otherwise no elements <i>should</i> be added to the queue nor submitted but it is a soft condition</li>
+   *   <li>When {@link #DRAIN_REQUIRED_MASK} is set, the queue overflow</li>
+   * </ul>
+   *
+   * @param element the element to add
+   * @return a bitset of [{@link #DRAIN_REQUIRED_MASK}, {@link #QUEUE_UNWRITABLE_MASK}, {@link #QUEUE_WRITABLE_MASK}] flags
+   */
+  public int add(E element) {
+    if (wip.compareAndSet(0, 1)) {
+      if (!consumer.test(element)) {
+        overflow = element;
+        return DRAIN_REQUIRED_MASK | (wip.get() == highWaterMark ? QUEUE_UNWRITABLE_MASK : 0);
+      }
+      if (consume(1) == 0) {
+        return 0;
+      }
+      return drainLoop();
+    } else {
+      queue.add(element);
+      return wip.incrementAndGet() == highWaterMark ? QUEUE_UNWRITABLE_MASK : 0;
```

Comment:
Mmmm I should read it again where...why we don't drain here if we win the race?
Between the compareAndSet and the subsequent incrementAndGet the would could have change enough that we are still eligible for becoming the one to drain...

Reply:
winning is implied by wip 0 -> 1 transition

Reply:
Yeah but if incrementAndGet == 0 it still means you have won....you can win in both ways (via compare and set or incrementAndGet) no?

Reply:
Actually here we just signal back-pressure to the consumer (it crossed the high-water mark), not related to the queue ownership.

Reply:
Yeah but if we have been the one to win it (by accident), no others will be able to take ownership anymore, unless fully drained. But draining happen only if someone win the race...

Reply:
ah right, this should be retried we can miss the signal here, good catch!!!

Reply:
I see what you mean but actually I don't think so because only the event-loop thread is allowed

- to call this method
- drain the queue (decrease the wip)

so if we don't get the ownership, it means another thread (producer) increased it and the value is greater than 1, consequently we cannot get ownership at this point.

Reply:
Nopez because these are two separate atomic ops and the world can change in many ways between them. If you fail the first cas(0, 1) you can succeed the next increment to still perform the same transition (0->1). 
Meaning that no others can perform it again and get the chance to drain, because there were no other concurrent requests while it has happened.

Reply:
ah so you mean that we can fail the first cas(0, 1) and wip still = 0 ?

Reply:
so

```java
      queue.add(element);
      long v = WIP_UPDATER.incrementAndGet(this);
      if (v == 1) {
        return drainLoop();
      } else {
        return v == highWaterMark ? QUEUE_UNWRITABLE_MASK : 0;
      }
```

would be correct then

Reply:
Exactly. The attempt for the compareAndSet(0, 1) is to avoid using the queue,. While te subsequent one can still fall in the same transition but the queue submission happens before (literally). In both cases draining is still required